File: structural/composite/composite.java
// component
interface Node {
    public String getName();
}

// leaf
class File implements Node {
    private String name;
    @Override
    public String getName(){ 
        return name;
    }
}

// composite
class Directory implements Node {
    private String name;
    private List<Node> children;
    @Override
    public String getName(){ 
        return name;
    }
    public void add(Node node) {
        children.add(node);
    }
}


// main
public class Main {
    public static void main(String[] args) 
    {
        Directory dir = new Directory();
        dir.add(new File());
        dir.add(new File());
        Directory secDir = new Directory();
        secDir.add(new File())
        dir.add(secDir);
    }
}

File: structural/composite/other_examples/composite.py
import abc


class Node(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def name(self):
        pass


class Directory(Node):
    def __init__(self, dirname):
        self.dirname = dirname
        self._children = set()

    def name(self):
        for child in self._children:
            child.name()
        print(self.dirname)

    def add(self, component):
        self._children.add(component)

    def remove(self, component):
        self._children.discard(component)


class File(Node):
    def __init__(self, filename):
        self.filename = filename

    def name(self):
        print(self.filename)


def main():
    dir = Directory('dir1')
    dir.add(File('file1'))
    dir.add(File('file2'))

    dir2 = Directory('dir2')
    dir2.add(File('file3'))
    dir.add(dir2)

    dir.name()


if __name__ == "__main__":
    main()
